api/src/routers/meals.js

```diff
@@ -0,0 +1,150 @@
+import express from "express";
+import knex from "../database_client.js";
+
+const mealsRouter = express.Router();
+
+mealsRouter.get("/", async (req, res, next) => {
+  try {
+    const query = knex("meal");
+    const {
+      maxPrice,
+      availableReservations,
+      title,
+      dateAfter,
+      dateBefore,
+      limit,
+      sortKey,
+      sortDir,
+    } = req.query;
+
+    if (maxPrice !== undefined) {
+      query.where("price", "<", maxPrice);
+    }
+    if (availableReservations !== undefined) {
+      if (availableReservations === "true") {
+        //might be returned as a string, hence checking explicitly
+        query
+          .leftJoin("reservation", "meal.id", "=", "reservation.meal_id")
+          .select("meal.id", "meal.max_reservations", "meal.title")
+          .sum("reservation.number_of_guests as sum_of_guests")
+          .groupBy("meal.id", "meal.max_reservations", "meal.title")
+          .having("sum_of_guests", "<", knex.ref("meal.max_reservations"));
+      } else {
+        query
+          .leftJoin("reservation", "meal.id", "=", "reservation.meal_id")
+          .select("meal.id", "meal.max_reservations", "meal.title")
+          .sum("reservation.number_of_guests as sum_of_guests")
+          .groupBy("meal.id", "meal.max_reservations", "meal.title")
+          .having("sum_of_guests", ">=", knex.ref("meal.max_reservations"));
+      }
+    }
+    if (title !== undefined) {
+      query.where("title", "like", `%${title}%`); //performing a partial match here
+    }
+    if (dateAfter !== undefined) {
+      query.where("when", ">", dateAfter);
+    }
+    if (dateBefore !== undefined) {
+      query.where("when", "<", dateBefore);
+    }
+    if (limit !== undefined) {
+      query.limit(limit); //Returns the given number of meals
+    }
```

Comment:
you are checking if the values in the request are undefined, but what if they are not, but incorrect data types? Validate the values and then return proper responses to the user so they know how to formulate the request correctly :)